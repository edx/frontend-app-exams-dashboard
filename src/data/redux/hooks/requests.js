import { useSelector, useDispatch } from 'react-redux';

import * as redux from 'data/redux';
import * as module from './requests'; // eslint-disable-line import/no-self-import

const selectors = redux.selectors.requests;
const actions = redux.actions.requests;

export const statusSelector = selector => (requestName) => useSelector(selector(requestName));
export const useRequestIsPending = module.statusSelector(selectors.isPending);
export const useRequestIsFailed = module.statusSelector(selectors.isFailed);
export const useRequestIsCompleted = module.statusSelector(selectors.isCompleted);
export const useRequestIsInactive = module.statusSelector(selectors.isInactive);
export const useRequestError = module.statusSelector(selectors.error);
export const useRequestErrorCode = module.statusSelector(selectors.errorCode);
export const useRequestErrorStatus = module.statusSelector(selectors.errorStatus);
export const useRequestData = module.statusSelector(selectors.data);

export const useMakeNetworkRequest = () => {
  const dispatch = useDispatch();
  return ({
    requestKey,
    promise,
    onSuccess,
    onFailure,
  }) => {
    dispatch(actions.startRequest({ requestKey }));
    return promise.then((response) => {
      if (onSuccess) { onSuccess(response); }
      dispatch(actions.completeRequest({ requestKey, response }));
    }).catch((error) => {
      // Parse backend error from axios object
      let errorParsed = error;
      if (error?.customAttributes?.httpErrorResponseData) {
        errorParsed = JSON.parse(error.customAttributes.httpErrorResponseData);
      }
      if (onFailure) { onFailure(errorParsed); }
      dispatch(actions.failRequest({ requestKey, error: errorParsed }));
    });
  };
};

export const useClearRequest = (requestKey) => {
  const dispatch = useDispatch();
  return () => {
    dispatch(actions.clearRequest({ requestKey }));
  };
};
