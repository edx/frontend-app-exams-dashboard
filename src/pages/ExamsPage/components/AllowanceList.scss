.allowances-header {
  padding: 40px 0 32px;
  display: flex;
  justify-content: space-between;
  gap: 10px;
}

.allowances-body {
  display: inline-block;
  text-align: center;
  width: 100%;
  padding-bottom: 50px;
}

.allowances-empty {
  padding: 100px 20px;
}

