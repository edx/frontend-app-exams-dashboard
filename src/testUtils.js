/* eslint-disable import/prefer-default-export */

/**
 * Mocked formatMessage provided by react-intl
 */
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) {
    return message;
  }
  // check if value is not a primitive type.
  if (Object.values(values).filter(value => Object(value) === value).length) {
    // eslint-disable-next-line react/jsx-filename-extension
    return <format-message-function {...{ message: msg, values }} />;
  }
  Object.keys(values).forEach((key) => {
    // eslint-disable-next-line
    message = message.replaceAll(`{${key}}`, values[key]);
  });
  return message;
};

export const defaultExamsData = {
  examsList: [
    {
      id: 1,
      name: 'exam1',
      examType: 'proctored',
      timeLimitMins: 60,
    },
  ],
  currentExamIndex: 0,
  currentExam: { id: 1, name: 'exam1' },
  setCurrentExam: jest.fn(),
};

export const defaultAttemptsData = {
  attemptsList:
    [{
      exam_name: 'Exam 1',
      username: 'username1',
      time_limit: 60,
      exam_type: 'timed',
      started_at: '2023-04-05T19:27:16.000000Z',
      completed_at: '2023-04-05T19:27:17.000000Z',
      status: 'second_review_required',
      attempt_id: 0,
      severity: 1.0,
      submission_reason: 'Submitted by user',
    },
    {
      exam_name: 'Exam 2',
      username: 'username2',
      time_limit: 60,
      exam_type: 'proctored',
      started_at: '2023-04-05T19:37:16.000000Z',
      completed_at: '2023-04-05T19:37:17.000000Z',
      status: 'submitted',
      attempt_id: 1,
    }],
};
