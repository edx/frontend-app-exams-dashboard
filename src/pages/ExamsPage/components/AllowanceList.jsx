import {
  Button,
  Collapsible,
  DataTable,
  useToggle,
} from '@openedx/paragon';
import { Add } from '@openedx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import messages from '../messages';
import { useAllowancesData } from '../hooks';
import AddAllowanceModal from './AddAllowanceModal';
import AllowanceListActions from './AllowanceListActions';

import './AllowanceList.scss';

const AllowanceList = () => {
  const { formatMessage } = useIntl();
  const [isOpen, open, close] = useToggle(false);

  const {
    allowancesList,
  } = useAllowancesData();

  const allowancesByUser = allowancesList.reduce((acc, curr) => {
    const {
      id,
      exam_id: examId,
      user_id: userId,
      extra_time_mins: extraTimeMins,
      username,
      exam_name: examName,
      email,
    } = curr;

    const allowance = {
      id,
      examId,
      examName,
      allowanceType: formatMessage(messages.allowanceTypeMinutes),
      extraTimeMins,
    };

    const user = acc.find(u => u.userId === userId);

    if (user) {
      user.allowances.push(allowance);

      return acc;
    }

    acc.push({
      userId,
      username,
      email,
      allowances: [allowance],
    });

    return acc;
  }, []);

  return (
    <>
      <div className="allowances-header">
        <h3 data-testid="allowances">{formatMessage(messages.allowanceDashboardTabTitle)}</h3>
        <Button
          variant="outline-primary"
          iconBefore={Add}
          size="sm"
          className="header-allowance-button"
          onClick={open}
        >
          {formatMessage(messages.addAllowanceButton)}
        </Button>
      </div>
      <div className="allowances-body">
        {!allowancesByUser.length
          ? (
            <div className="allowances-empty">
              <h4>{formatMessage(messages.noAllowancesHeader)}</h4>
              <p>{formatMessage(messages.noAllowancesBody)}</p>
              <Button variant="primary" iconBefore={Add} onClick={open}>
                {formatMessage(messages.addAllowanceButton)}
              </Button>
            </div>
          ) : (
            <div className="allowances-list">
              {allowancesByUser.map(({
                userId, username, email, allowances,
              }) => (
                <Collapsible
                  defaultOpen
                  key={userId}
                  styling="card-lg"
                  title={(
                    <p>
                      <strong>{username}</strong> (<a href={`mailto: ${email}`} target="_blank" onClick={e => e.stopPropagation()} rel="noreferrer">{email}</a>)
                    </p>
                  )}
                >
                  <span />
                  <DataTable
                    itemCount={allowances.length}
                    data={allowances}
                    columns={[
                      {
                        Header: formatMessage(messages.tableColumnExamName),
                        accessor: 'examName',

                      },
                      {
                        Header: formatMessage(messages.tableColumnAllowanceType),
                        accessor: 'allowanceType',
                      },
                      {
                        Header: formatMessage(messages.tableColumnAllowanceValue),
                        accessor: 'extraTimeMins',
                      },
                    ]}
                    additionalColumns={[
                      {
                        id: 'actions',
                        Cell: ({ row }) => AllowanceListActions({ allowance: row.original }),
                      },
                    ]}
                  >
                    <DataTable.Table />
                  </DataTable>
                </Collapsible>
              ))}
            </div>
          )}
      </div>
      <AddAllowanceModal isOpen={isOpen} close={close} />
    </>
  );
};

export default AllowanceList;
