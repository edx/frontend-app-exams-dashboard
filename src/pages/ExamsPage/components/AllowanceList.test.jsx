import { render, screen } from '@testing-library/react';

import AllowanceList from './AllowanceList';
import * as hooks from '../hooks';
import * as testUtils from '../../../testUtils';

// normally mocked for unit tests but required for rendering/snapshots
jest.unmock('react');

const mockMakeNetworkRequest = jest.fn();

const mockedAllowancesData = {
  allowancesList: [
    {
      id: 1,
      exam_id: 1,
      user_id: 1,
      extra_time_mins: 45,
      username: 'edx',
      exam_name: 'edX Exams',
      email: 'edx@example.com',
    },
    {
      id: 2,
      exam_id: 3,
      user_id: 1,
      extra_time_mins: 15,
      username: 'edx',
      exam_name: 'More Ways to Connect',
      email: 'edx@example.com',
    },
  ],
};

jest.mock('../hooks', () => ({
  useAllowancesData: jest.fn(),
  useExamsData: jest.fn(),
  useButtonStateFromRequestStatus: jest.fn(),
  useCreateAllowance: jest.fn(),
  useFilteredExamsData: jest.fn(),
  useEditAllowance: jest.fn(),
  useDeleteAllowance: jest.fn(),
}));

jest.mock('../../../data/redux/hooks', () => ({
  useRequestError: jest.fn(),
  useClearRequest: jest.fn(),
}));

describe('AllowanceList', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    hooks.useExamsData.mockReturnValue(testUtils.defaultExamsData);
    hooks.useButtonStateFromRequestStatus.mockReturnValue(mockMakeNetworkRequest);
    hooks.useFilteredExamsData.mockReturnValue({ proctoredExams: {}, timedExams: {} });
  });

  describe('when listing allowances', () => {
    beforeEach(() => {
      hooks.useAllowancesData.mockReturnValue(mockedAllowancesData);
    });

    it('should match snapshot', () => {
      expect(render(<AllowanceList />)).toMatchSnapshot();
    });

    it('should open allowance modal', () => {
      render(<AllowanceList />);
      screen.getByText('Add allowance').click();
      expect(screen.getByText('Add a new allowance')).toBeInTheDocument();
    });
  });

  describe('when there are no allowances', () => {
    beforeEach(() => {
      hooks.useAllowancesData.mockReturnValue({ allowancesList: [] });
    });

    it('should match snapshot', () => {
      expect(render(<AllowanceList />)).toMatchSnapshot();
    });
  });
});
