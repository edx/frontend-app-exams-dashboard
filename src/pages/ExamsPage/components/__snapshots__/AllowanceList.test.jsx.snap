// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllowanceList when listing allowances should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="allowances-header"
      >
        <h3
          data-testid="allowances"
        >
          Allowances
        </h3>
        <button
          class="header-allowance-button btn btn-outline-primary btn-sm"
          type="button"
        >
          <span
            class="pgn__icon pgn__icon__sm btn-icon-before"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
                fill="currentColor"
              />
            </svg>
          </span>
          Add allowance
        </button>
      </div>
      <div
        class="allowances-body"
      >
        <div
          class="allowances-list"
        >
          <div
            class="pgn_collapsible collapsible-card-lg is-open"
          >
            <div
              aria-expanded="true"
              class="collapsible-trigger"
              role="button"
              tabindex="0"
            >
              <p>
                <strong>
                  edx
                </strong>
                 (
                <a
                  href="mailto: edx@example.com"
                  rel="noreferrer"
                  target="_blank"
                >
                  edx@example.com
                </a>
                )
              </p>
              <span
                class="collapsible-icon"
              >
                <span
                  class="pgn__icon"
                >
                  <svg
                    aria-hidden="true"
                    fill="none"
                    focusable="false"
                    height="24"
                    role="img"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14l-6-6Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              class="pgn-transition-replace-group position-relative"
            >
              <div
                style="padding: .1px 0px;"
              >
                <div
                  class="collapsible-body"
                >
                  <span />
                  <div
                    class="pgn__data-table-layout-wrapper"
                  >
                    <div
                      class="pgn__data-table-layout-main"
                    >
                      <div
                        class="pgn__data-table-wrapper"
                      >
                        <div
                          class="pgn__data-table-container"
                          data-testid="data-table-container"
                        >
                          <table
                            class="pgn__data-table is-striped"
                            role="table"
                          >
                            <thead>
                              <tr
                                role="row"
                              >
                                <th
                                  colspan="1"
                                  role="columnheader"
                                >
                                  <span
                                    class="d-flex align-items-center"
                                  >
                                    <span>
                                      Exam name
                                    </span>
                                  </span>
                                </th>
                                <th
                                  colspan="1"
                                  role="columnheader"
                                >
                                  <span
                                    class="d-flex align-items-center"
                                  >
                                    <span>
                                      Allowance type
                                    </span>
                                  </span>
                                </th>
                                <th
                                  colspan="1"
                                  role="columnheader"
                                >
                                  <span
                                    class="d-flex align-items-center"
                                  >
                                    <span>
                                      Allowance value
                                    </span>
                                  </span>
                                </th>
                                <th
                                  colspan="1"
                                  role="columnheader"
                                >
                                  <span
                                    class="d-flex align-items-center"
                                  >
                                    <span>
                                      Â 
                                    </span>
                                  </span>
                                </th>
                              </tr>
                            </thead>
                            <tbody
                              role="rowgroup"
                            >
                              <tr
                                class="pgn__data-table-row"
                                role="row"
                              >
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  edX Exams
                                </td>
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  Additional time (minutes)
                                </td>
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  45
                                </td>
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  <div
                                    class="allowances-actions text-right"
                                  >
                                    <button
                                      aria-label="Edit allowance"
                                      class="btn-icon btn-icon-primary btn-icon-sm mr-2"
                                      data-testid="edit-allowance-icon"
                                      type="button"
                                    >
                                      <span
                                        class="btn-icon__icon-container"
                                      >
                                        <span
                                          class="pgn__icon btn-icon__icon"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            fill="none"
                                            focusable="false"
                                            height="24"
                                            role="img"
                                            viewBox="0 0 24 24"
                                            width="24"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <path
                                              d="m14.06 9.02.92.92L5.92 19H5v-.92l9.06-9.06ZM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 0 0 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29Zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75Z"
                                              fill="currentColor"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </button>
                                    <button
                                      aria-label="Delete allowance"
                                      class="btn-icon btn-icon-secondary btn-icon-sm"
                                      data-testid="delete-allowance-icon"
                                      type="button"
                                    >
                                      <span
                                        class="btn-icon__icon-container"
                                      >
                                        <span
                                          class="pgn__icon btn-icon__icon"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            fill="none"
                                            focusable="false"
                                            height="24"
                                            role="img"
                                            viewBox="0 0 24 24"
                                            width="24"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <path
                                              d="M6 21h12V7H6v14zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4h-3.5z"
                                              fill="currentColor"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </button>
                                  </div>
                                </td>
                              </tr>
                              <tr
                                class="pgn__data-table-row"
                                role="row"
                              >
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  More Ways to Connect
                                </td>
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  Additional time (minutes)
                                </td>
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  15
                                </td>
                                <td
                                  class="pgn__data-table-cell-wrap"
                                  role="cell"
                                >
                                  <div
                                    class="allowances-actions text-right"
                                  >
                                    <button
                                      aria-label="Edit allowance"
                                      class="btn-icon btn-icon-primary btn-icon-sm mr-2"
                                      data-testid="edit-allowance-icon"
                                      type="button"
                                    >
                                      <span
                                        class="btn-icon__icon-container"
                                      >
                                        <span
                                          class="pgn__icon btn-icon__icon"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            fill="none"
                                            focusable="false"
                                            height="24"
                                            role="img"
                                            viewBox="0 0 24 24"
                                            width="24"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <path
                                              d="m14.06 9.02.92.92L5.92 19H5v-.92l9.06-9.06ZM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 0 0 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29Zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75Z"
                                              fill="currentColor"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </button>
                                    <button
                                      aria-label="Delete allowance"
                                      class="btn-icon btn-icon-secondary btn-icon-sm"
                                      data-testid="delete-allowance-icon"
                                      type="button"
                                    >
                                      <span
                                        class="btn-icon__icon-container"
                                      >
                                        <span
                                          class="pgn__icon btn-icon__icon"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            fill="none"
                                            focusable="false"
                                            height="24"
                                            role="img"
                                            viewBox="0 0 24 24"
                                            width="24"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <path
                                              d="M6 21h12V7H6v14zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4h-3.5z"
                                              fill="currentColor"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </button>
                                  </div>
                                </td>
                              </tr>
                            </tbody>
                          </table>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="allowances-header"
    >
      <h3
        data-testid="allowances"
      >
        Allowances
      </h3>
      <button
        class="header-allowance-button btn btn-outline-primary btn-sm"
        type="button"
      >
        <span
          class="pgn__icon pgn__icon__sm btn-icon-before"
        >
          <svg
            aria-hidden="true"
            fill="none"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
              fill="currentColor"
            />
          </svg>
        </span>
        Add allowance
      </button>
    </div>
    <div
      class="allowances-body"
    >
      <div
        class="allowances-list"
      >
        <div
          class="pgn_collapsible collapsible-card-lg is-open"
        >
          <div
            aria-expanded="true"
            class="collapsible-trigger"
            role="button"
            tabindex="0"
          >
            <p>
              <strong>
                edx
              </strong>
               (
              <a
                href="mailto: edx@example.com"
                rel="noreferrer"
                target="_blank"
              >
                edx@example.com
              </a>
              )
            </p>
            <span
              class="collapsible-icon"
            >
              <span
                class="pgn__icon"
              >
                <svg
                  aria-hidden="true"
                  fill="none"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14l-6-6Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </span>
          </div>
          <div
            class="pgn-transition-replace-group position-relative"
          >
            <div
              style="padding: .1px 0px;"
            >
              <div
                class="collapsible-body"
              >
                <span />
                <div
                  class="pgn__data-table-layout-wrapper"
                >
                  <div
                    class="pgn__data-table-layout-main"
                  >
                    <div
                      class="pgn__data-table-wrapper"
                    >
                      <div
                        class="pgn__data-table-container"
                        data-testid="data-table-container"
                      >
                        <table
                          class="pgn__data-table is-striped"
                          role="table"
                        >
                          <thead>
                            <tr
                              role="row"
                            >
                              <th
                                colspan="1"
                                role="columnheader"
                              >
                                <span
                                  class="d-flex align-items-center"
                                >
                                  <span>
                                    Exam name
                                  </span>
                                </span>
                              </th>
                              <th
                                colspan="1"
                                role="columnheader"
                              >
                                <span
                                  class="d-flex align-items-center"
                                >
                                  <span>
                                    Allowance type
                                  </span>
                                </span>
                              </th>
                              <th
                                colspan="1"
                                role="columnheader"
                              >
                                <span
                                  class="d-flex align-items-center"
                                >
                                  <span>
                                    Allowance value
                                  </span>
                                </span>
                              </th>
                              <th
                                colspan="1"
                                role="columnheader"
                              >
                                <span
                                  class="d-flex align-items-center"
                                >
                                  <span>
                                    Â 
                                  </span>
                                </span>
                              </th>
                            </tr>
                          </thead>
                          <tbody
                            role="rowgroup"
                          >
                            <tr
                              class="pgn__data-table-row"
                              role="row"
                            >
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                edX Exams
                              </td>
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                Additional time (minutes)
                              </td>
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                45
                              </td>
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                <div
                                  class="allowances-actions text-right"
                                >
                                  <button
                                    aria-label="Edit allowance"
                                    class="btn-icon btn-icon-primary btn-icon-sm mr-2"
                                    data-testid="edit-allowance-icon"
                                    type="button"
                                  >
                                    <span
                                      class="btn-icon__icon-container"
                                    >
                                      <span
                                        class="pgn__icon btn-icon__icon"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          fill="none"
                                          focusable="false"
                                          height="24"
                                          role="img"
                                          viewBox="0 0 24 24"
                                          width="24"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <path
                                            d="m14.06 9.02.92.92L5.92 19H5v-.92l9.06-9.06ZM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 0 0 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29Zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75Z"
                                            fill="currentColor"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </button>
                                  <button
                                    aria-label="Delete allowance"
                                    class="btn-icon btn-icon-secondary btn-icon-sm"
                                    data-testid="delete-allowance-icon"
                                    type="button"
                                  >
                                    <span
                                      class="btn-icon__icon-container"
                                    >
                                      <span
                                        class="pgn__icon btn-icon__icon"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          fill="none"
                                          focusable="false"
                                          height="24"
                                          role="img"
                                          viewBox="0 0 24 24"
                                          width="24"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <path
                                            d="M6 21h12V7H6v14zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4h-3.5z"
                                            fill="currentColor"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </button>
                                </div>
                              </td>
                            </tr>
                            <tr
                              class="pgn__data-table-row"
                              role="row"
                            >
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                More Ways to Connect
                              </td>
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                Additional time (minutes)
                              </td>
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                15
                              </td>
                              <td
                                class="pgn__data-table-cell-wrap"
                                role="cell"
                              >
                                <div
                                  class="allowances-actions text-right"
                                >
                                  <button
                                    aria-label="Edit allowance"
                                    class="btn-icon btn-icon-primary btn-icon-sm mr-2"
                                    data-testid="edit-allowance-icon"
                                    type="button"
                                  >
                                    <span
                                      class="btn-icon__icon-container"
                                    >
                                      <span
                                        class="pgn__icon btn-icon__icon"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          fill="none"
                                          focusable="false"
                                          height="24"
                                          role="img"
                                          viewBox="0 0 24 24"
                                          width="24"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <path
                                            d="m14.06 9.02.92.92L5.92 19H5v-.92l9.06-9.06ZM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83a.996.996 0 0 0 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29Zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75Z"
                                            fill="currentColor"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </button>
                                  <button
                                    aria-label="Delete allowance"
                                    class="btn-icon btn-icon-secondary btn-icon-sm"
                                    data-testid="delete-allowance-icon"
                                    type="button"
                                  >
                                    <span
                                      class="btn-icon__icon-container"
                                    >
                                      <span
                                        class="pgn__icon btn-icon__icon"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          fill="none"
                                          focusable="false"
                                          height="24"
                                          role="img"
                                          viewBox="0 0 24 24"
                                          width="24"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <path
                                            d="M6 21h12V7H6v14zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4h-3.5z"
                                            fill="currentColor"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </button>
                                </div>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AllowanceList when there are no allowances should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class=""
  >
    <div
      id="paragon-portal-root"
    />
    <div>
      <div
        class="allowances-header"
      >
        <h3
          data-testid="allowances"
        >
          Allowances
        </h3>
        <button
          class="header-allowance-button btn btn-outline-primary btn-sm"
          type="button"
        >
          <span
            class="pgn__icon pgn__icon__sm btn-icon-before"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
                fill="currentColor"
              />
            </svg>
          </span>
          Add allowance
        </button>
      </div>
      <div
        class="allowances-body"
      >
        <div
          class="allowances-empty"
        >
          <h4>
            No Allowances
          </h4>
          <p>
            Need to grant an allowance? Get started here.
          </p>
          <button
            class="btn btn-primary"
            type="button"
          >
            <span
              class="pgn__icon btn-icon-before"
            >
              <svg
                aria-hidden="true"
                fill="none"
                focusable="false"
                height="24"
                role="img"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            Add allowance
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="allowances-header"
    >
      <h3
        data-testid="allowances"
      >
        Allowances
      </h3>
      <button
        class="header-allowance-button btn btn-outline-primary btn-sm"
        type="button"
      >
        <span
          class="pgn__icon pgn__icon__sm btn-icon-before"
        >
          <svg
            aria-hidden="true"
            fill="none"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
              fill="currentColor"
            />
          </svg>
        </span>
        Add allowance
      </button>
    </div>
    <div
      class="allowances-body"
    >
      <div
        class="allowances-empty"
      >
        <h4>
          No Allowances
        </h4>
        <p>
          Need to grant an allowance? Get started here.
        </p>
        <button
          class="btn btn-primary"
          type="button"
        >
          <span
            class="pgn__icon btn-icon-before"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="24"
              role="img"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
                fill="currentColor"
              />
            </svg>
          </span>
          Add allowance
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
